### PR TITLE
CASSANDRA-13704 Safer handling of out-of-range tokens

### PR DESCRIPTION
**Pull request by maedhroz**
patch by Caleb Rackliffe; reviewed by ? for [CASSANDRA-13704](https://issues.apache.org/jira/browse/CASSANDRA-13704)

Co-authored-by: Sam Tunnicliffe [sam@beobal.com](mailto:sam@beobal.com)
Co-authored-by: Caleb Rackliffe [calebrackliffe@gmail.com](mailto:calebrackliffe@gmail.com)

### CODE REVIEW

File: src/java/org/apache/cassandra/service/StorageService.java [hunk 1]
```diff
@@ -349,6 +350,35 @@ private static int getSchemaDelay()
     @VisibleForTesting // this is used for dtests only, see CASSANDRA-18152
     public volatile boolean skipNotificationListeners = false;
 
+    private final java.util.function.Predicate<Keyspace> anyOutOfRangeOpsRecorded = 
+                keyspace -> keyspace.metric.outOfRangeTokenReads.getCount() > 0 ||
+                            keyspace.metric.outOfRangeTokenWrites.getCount() > 0 ||
+                    keyspace.metric.outOfRangeTokenPaxosRequests.getCount() > 0;
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/service/StorageService.java]: TODO: rearrange `||` s

File: src/java/org/apache/cassandra/dht/OwnedRanges.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+import org.apache.cassandra.locator.InetAddressAndPort;
+import org.apache.cassandra.metrics.StorageMetrics;
+import org.apache.cassandra.service.StorageService;
+
+public final class OwnedRanges
+{
+    private static final Logger logger = LoggerFactory.getLogger(OwnedRanges.class);
+
+    private static final Comparator<Range<Token>> rangeComparator = (r1, r2) ->
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/dht/OwnedRanges.java:38]: TODO: comparator chain

File: src/java/org/apache/cassandra/dht/OwnedRanges.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                logger.warn("[{}] Received {} from {} containing ranges {} outside valid ranges {}",
+                            requestId, requestType, from, unownedRanges, ownedRanges);
+            }
+        }
+
+        return !outOfRangeTokenRejection || unownedRanges.isEmpty();
+    }
+
+    /**
+     * Takes a collection of ranges and returns ranges from that collection that are not covered by the this node's owned ranges.
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/dht/OwnedRanges.java]:
```suggestion
     * Takes a collection of ranges and returns ranges from that collection that are not covered by this node's owned ranges.
```

File: src/java/org/apache/cassandra/dht/OwnedRanges.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+     * If either option is enabled and we do detect unowned ranges in the request, we increment a metric then take further
+     * action depending on the config.
+     *
+     * @param requestedRanges the set of token ranges contained in a request from a peer
+     * @param requestId an identifier for the peer request, to be used in logging (e.g. Stream or Repair Session #)
+     * @param requestType description of the request type, to be used in logging (e.g. "prepare request" or "validation")
+     * @param from the originator of the request
+     * @return true if the request should be accepted (either because no checking was performed, invalid ranges were d
+     *         identified but only the logging action is enabled, or because all request ranges were valid. Otherwise,
+     *         returns false to indicate the request should be rejected.
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/dht/OwnedRanges.java]: see https://github.com/apache/cassandra/pull/3539/files#r1759388209

**Comment by maedhroz** [comment]
https://ci-cassandra.apache.org/view/patches/job/Cassandra-devbranch-5/64/

**Comment by maedhroz** [comment]
https://ci-cassandra.apache.org/view/patches/job/Cassandra-devbranch-5/67/

**Review by michaelsembwever — APPROVED** [review]

**Comment by maedhroz** [comment]
Committed as https://github.com/apache/cassandra/commit/2c94b27bedb43da0069b97b2211802f4deadc071

**Pull request closed without merging**